**About Dataset**

**Context**

"Predict behavior to retain customers. You can analyze all relevant customer data and develop focused customer retention programs." [IBM Sample Data Sets]

**Content**

Each row represents a customer, each column contains customer’s attributes described on the column Metadata.

**The data set includes information about:**

Customers who left within the last month – the column is called Churn

Services that each customer has signed up for – phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies

Customer account information – how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges
Demographic info about customers – gender, age range, and if they have partners and dependents

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
from google.colab import files
uploaded = files.upload()

Saving customer_churn.csv to customer_churn.csv


In [3]:
data = pd.read_csv("customer_churn.csv")
data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
data.shape

(7043, 21)

In [5]:
data.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [6]:
data.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [7]:
list(data.columns)

['customerID',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'MonthlyCharges',
 'TotalCharges',
 'Churn']

In [8]:
data.nunique()

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

In [9]:
uni_dict = {}

for col in data.columns:
  uni_dict[col] = [data[col].unique()]

In [10]:
uni_dict

{'customerID': [array(['7590-VHVEG', '5575-GNVDE', '3668-QPYBK', ..., '4801-JZAZL',
         '8361-LTMKD', '3186-AJIEK'], dtype=object)],
 'gender': [array(['Female', 'Male'], dtype=object)],
 'SeniorCitizen': [array([0, 1])],
 'Partner': [array(['Yes', 'No'], dtype=object)],
 'Dependents': [array(['No', 'Yes'], dtype=object)],
 'tenure': [array([ 1, 34,  2, 45,  8, 22, 10, 28, 62, 13, 16, 58, 49, 25, 69, 52, 71,
         21, 12, 30, 47, 72, 17, 27,  5, 46, 11, 70, 63, 43, 15, 60, 18, 66,
          9,  3, 31, 50, 64, 56,  7, 42, 35, 48, 29, 65, 38, 68, 32, 55, 37,
         36, 41,  6,  4, 33, 67, 23, 57, 61, 14, 20, 53, 40, 59, 24, 44, 19,
         54, 51, 26,  0, 39])],
 'PhoneService': [array(['No', 'Yes'], dtype=object)],
 'MultipleLines': [array(['No phone service', 'No', 'Yes'], dtype=object)],
 'InternetService': [array(['DSL', 'Fiber optic', 'No'], dtype=object)],
 'OnlineSecurity': [array(['No', 'Yes', 'No internet service'], dtype=object)],
 'OnlineBackup': [array(['Yes', 'No'

1]totalCharges is object convert to float
2]in many column for 'No' there are two different values

In [11]:
data["TotalCharges"] = pd.to_numeric(data["TotalCharges"], errors="coerce")
data["TotalCharges"].fillna(data["TotalCharges"].median(), inplace=True)


In [12]:
data["MultipleLines"] = data["MultipleLines"].replace("No phone service", "No")

internet_cols = ["OnlineSecurity", "OnlineBackup", "DeviceProtection",
                 "TechSupport", "StreamingTV", "StreamingMovies"]
for col in internet_cols:
    data[col] = data[col].replace("No internet service", "No")


Convert 'Yes' and 'No' to 1 and 0

In [13]:
binary_cols = ['Churn', 'Partner', 'Dependents', 'PhoneService','MultipleLines' ,  'PaperlessBilling', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']

for col in binary_cols:
  data[col] = data[col].replace({'Yes':1,'No':0})

SeniorCitizen columns is already in 1/0 so make sure i should in 'int'

In [14]:
data['SeniorCitizen'] = data['SeniorCitizen'].astype(int)

In [15]:
data.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG  Female              0        1           0       1   
1  5575-GNVDE    Male              0        0           0      34   
2  3668-QPYBK    Male              0        0           0       2   
3  7795-CFOCW    Male              0        0           0      45   
4  9237-HQITU  Female              0        0           0       2   

   PhoneService  MultipleLines InternetService  OnlineSecurity  ...  \
0             0              0             DSL               0  ...   
1             1              0             DSL               1  ...   
2             1              0             DSL               1  ...   
3             0              0             DSL               1  ...   
4             1              0     Fiber optic               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
0                 0            0            0                0   
1                 1            0            0                0   
2                 0            0            0                0   
3                 1            1            0                0   
4                 0            0            0                0   

         Contract PaperlessBilling              PaymentMethod MonthlyCharges  \
0  Month-to-month                1           Electronic check          29.85   
1        One year                0               Mailed check          56.95   
2  Month-to-month                1               Mailed check          53.85   
3        One year                0  Bank transfer (automatic)          42.30   
4  Month-to-month                1           Electronic check          70.70   

   TotalCharges  Churn  
0         29.85      0  
1       1889.50      0  
2        108.15      1  
3       1840.75      0  
4        151.65      1  

[5 rows x 21 columns]

apply get_dummies on remaining categorical columns

In [16]:
cat_col = ['gender', 'PaymentMethod', 'Contract', 'InternetService']

for col in cat_col:
  data = pd.concat([data,pd.get_dummies(data[col],prefix=col ,drop_first=True)],axis =1)

data.drop(columns = cat_col,inplace =True , axis =1)
data.head()

customerID  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0  7590-VHVEG              0        1           0       1             0   
1  5575-GNVDE              0        0           0      34             1   
2  3668-QPYBK              0        0           0       2             1   
3  7795-CFOCW              0        0           0      45             0   
4  9237-HQITU              0        0           0       2             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
0              0               0             1                 0  ...   
1              0               1             0                 1  ...   
2              0               1             1                 0  ...   
3              0               1             0                 1  ...   
4              0               0             0                 0  ...   

   TotalCharges  Churn  gender_Male  PaymentMethod_Credit card (automatic)  \
0         29.85      0        False                                  False   
1       1889.50      0         True                                  False   
2        108.15      1         True                                  False   
3       1840.75      0         True                                  False   
4        151.65      1        False                                  False   

   PaymentMethod_Electronic check  PaymentMethod_Mailed check  \
0                            True                       False   
1                           False                        True   
2                           False                        True   
3                           False                       False   
4                            True                       False   

   Contract_One year  Contract_Two year  InternetService_Fiber optic  \
0              False              False                        False   
1               True              False                        False   
2              False              False                        False   
3               True              False                        False   
4              False              False                         True   

   InternetService_No  
0               False  
1               False  
2               False  
3               False  
4               False  

[5 rows x 25 columns]

In [17]:
data.replace({
    True:1,
    False:0
},inplace=True)

In [18]:
data.head()

customerID  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0  7590-VHVEG              0        1           0       1             0   
1  5575-GNVDE              0        0           0      34             1   
2  3668-QPYBK              0        0           0       2             1   
3  7795-CFOCW              0        0           0      45             0   
4  9237-HQITU              0        0           0       2             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
0              0               0             1                 0  ...   
1              0               1             0                 1  ...   
2              0               1             1                 0  ...   
3              0               1             0                 1  ...   
4              0               0             0                 0  ...   

   TotalCharges  Churn  gender_Male  PaymentMethod_Credit card (automatic)  \
0         29.85      0            0                                      0   
1       1889.50      0            1                                      0   
2        108.15      1            1                                      0   
3       1840.75      0            1                                      0   
4        151.65      1            0                                      0   

   PaymentMethod_Electronic check  PaymentMethod_Mailed check  \
0                               1                           0   
1                               0                           1   
2                               0                           1   
3                               0                           0   
4                               1                           0   

   Contract_One year  Contract_Two year  InternetService_Fiber optic  \
0                  0                  0                            0   
1                  1                  0                            0   
2                  0                  0                            0   
3                  1                  0                            0   
4                  0                  0                            1   

   InternetService_No  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 25 columns]

Normalize the continues type data containing columns

In [19]:
nor_col = ['tenure','MonthlyCharges','TotalCharges']
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
data[nor_col] = scaler.fit_transform(data[nor_col])
data.head()

customerID  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
0  7590-VHVEG              0        1           0  0.013889             0   
1  5575-GNVDE              0        0           0  0.472222             1   
2  3668-QPYBK              0        0           0  0.027778             1   
3  7795-CFOCW              0        0           0  0.625000             0   
4  9237-HQITU              0        0           0  0.027778             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  ...  \
0              0               0             1                 0  ...   
1              0               1             0                 1  ...   
2              0               1             1                 0  ...   
3              0               1             0                 1  ...   
4              0               0             0                 0  ...   

   TotalCharges  Churn  gender_Male  PaymentMethod_Credit card (automatic)  \
0      0.001275      0            0                                      0   
1      0.215867      0            1                                      0   
2      0.010310      1            1                                      0   
3      0.210241      0            1                                      0   
4      0.015330      1            0                                      0   

   PaymentMethod_Electronic check  PaymentMethod_Mailed check  \
0                               1                           0   
1                               0                           1   
2                               0                           1   
3                               0                           0   
4                               1                           0   

   Contract_One year  Contract_Two year  InternetService_Fiber optic  \
0                  0                  0                            0   
1                  1                  0                            0   
2                  0                  0                            0   
3                  1                  0                            0   
4                  0                  0                            1   

   InternetService_No  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 25 columns]

drop unwanted columns

In [20]:
data.drop(columns = ['customerID'],inplace=True,axis=1)

In [21]:
dt = data.copy()

Split dataset into Train and Test sets

In [22]:
from sklearn.model_selection import train_test_split
X = dt.drop(columns =['Churn'])
y = dt['Churn']

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [23]:
X_train.shape,y_train.shape

((5634, 23), (5634,))

In [24]:
X_train.shape[1]

23

In [25]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.optimizers import Adam

model = keras.Sequential([
    keras.layers.Dense(64,activation='relu',input_shape=(X_train.shape[1],)),
    keras.layers.Dense(32,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

optimizer = Adam(learning_rate=0.0005)
model.compile(
    optimizer=optimizer,
    loss='binary_crossentropy',
    metrics=['accuracy']
)

model.fit(X_train,y_train,epochs=50 , batch_size=32,validation_data=(X_test,y_test))

Epoch 1/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.7026 - loss: 0.5707 - val_accuracy: 0.8041 - val_loss: 0.4210
Epoch 2/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7934 - loss: 0.4386 - val_accuracy: 0.8126 - val_loss: 0.4071
Epoch 3/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7926 - loss: 0.4402 - val_accuracy: 0.8048 - val_loss: 0.4054
Epoch 4/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8054 - loss: 0.4219 - val_accuracy: 0.8176 - val_loss: 0.4025
Epoch 5/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8004 - loss: 0.4214 - val_accuracy: 0.8190 - val_loss: 0.4018
Epoch 6/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7962 - loss: 0.4193 - val_accuracy: 0.8204 - val_loss: 0.4022
Epoch 7/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7940 - loss: 0.4185 - val_accuracy: 0.8169 - val_loss: 0.4038
Epoch 8/50
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8020 - loss: 0.4228 - val_accuracy: 0.

In [26]:
X_train.shape

(5634, 23)

Balancing Data

In [35]:
import tensorflow as tf
from tensorflow import keras
from sklearn.metrics import confusion_matrix , classification_report
import numpy as np

def ANN(X_train,X_test,y_train,y_test,loss,weights):
  model = keras.Sequential([
      keras.layers.Dense(64,activation='relu',input_shape=(X_train.shape[1],)),
      keras.layers.Dense(32,activation='relu'),
      keras.layers.Dense(1,activation='sigmoid')
  ])

  model.compile(optimizer='adam',loss=loss,metrics=['accuracy'])

  if weights == -1:
    model.fit(X_train,y_train,epochs=100,validation_data=(X_test,y_test))
  else:
     model.fit(X_train,y_train,epochs=100,validation_data=(X_test,y_test),class_weight=weights)

  y_pred = model.predict(X_test)
  y_pred = np.round(y_pred)

  print("classification Report : \n",classification_report(y_test,y_pred))

  return y_pred


In [36]:
y_prediction = ANN(X_train,X_test,y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 36s 197ms/step - accuracy: 0.7507 - loss: 0.5184 - val_accuracy: 0.8155 - val_loss: 0.4069
Epoch 2/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8012 - loss: 0.4221 - val_accuracy: 0.8070 - val_loss: 0.4096
Epoch 3/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8016 - loss: 0.4207 - val_accuracy: 0.8183 - val_loss: 0.4005
Epoch 4/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8007 - loss: 0.4234 - val_accuracy: 0.8070 - val_loss: 0.4053
Epoch 5/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8052 - loss: 0.4083 - val_accuracy: 0.8197 - val_loss: 0.4000
Epoch 6/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8063 - loss: 0.4028 - val_accuracy: 0.8148 - val_loss: 0.3991
Epoch 7/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7978 - loss: 0.4251 - val_accuracy: 0.8133 - val_loss: 0.4012
Epoch 8/100
177/177 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8159 - loss: 0.3969 - val_a

Precision , recall , f1-score for the class_1 is low compared to Class_0 bcz of imbalanced Data .

In [38]:
 dt['Churn'].value_counts()

Churn
0    5174
1    1869
Name: count, dtype: int64

Balancing Data Using Under Sampling

In [39]:
count_0 , count_1 = dt['Churn'].value_counts()

In [41]:
class_0 = dt[dt['Churn']==0]
class_1 = dt[dt['Churn']==1]
class_0.shape

(5174, 24)

In [45]:
sample_class_0 = class_0.sample(count_1)
dt_under = pd.concat([sample_class_0,class_1],axis =0)
dt_under['Churn'].value_counts()

Churn
0    1869
1    1869
Name: count, dtype: int64

In [46]:
X = dt_under.drop(columns=['Churn'])
y = dt_under['Churn']

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state = 42)

y_prediction = ANN(X_train,X_test,y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 4s 22ms/step - accuracy: 0.6882 - loss: 0.6049 - val_accuracy: 0.7460 - val_loss: 0.4929
Epoch 2/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7588 - loss: 0.5030 - val_accuracy: 0.7553 - val_loss: 0.4876
Epoch 3/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7703 - loss: 0.4843 - val_accuracy: 0.7647 - val_loss: 0.4853
Epoch 4/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7780 - loss: 0.4717 - val_accuracy: 0.7553 - val_loss: 0.4853
Epoch 5/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7730 - loss: 0.4738 - val_accuracy: 0.7620 - val_loss: 0.4836
Epoch 6/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7632 - loss: 0.4816 - val_accuracy: 0.7607 - val_loss: 0.4872
Epoch 7/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7762 - loss: 0.4657 - val_accuracy: 0.7513 - val_loss: 0.4889
Epoch 8/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7803 - loss: 0.4657 - val_accuracy: 0.7594 - 

Using Undersampling F1_score for class_1 increasing from 0.52 to 0.72

Using Oversampling

In [47]:
class_1_over = class_1.sample(count_0 , replace =True)
dt_over = pd.concat([class_0,class_1_over],axis = 0)
dt_over['Churn'].value_counts()

Churn
0    5174
1    5174
Name: count, dtype: int64

In [48]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42,stratify=y)
y_prediction = ANN(X_train,X_test,y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 3s 11ms/step - accuracy: 0.6451 - loss: 0.6315 - val_accuracy: 0.7473 - val_loss: 0.5061
Epoch 2/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7678 - loss: 0.4973 - val_accuracy: 0.7594 - val_loss: 0.4924
Epoch 3/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7523 - loss: 0.5043 - val_accuracy: 0.7540 - val_loss: 0.4931
Epoch 4/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7655 - loss: 0.4905 - val_accuracy: 0.7647 - val_loss: 0.4857
Epoch 5/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7698 - loss: 0.4786 - val_accuracy: 0.7620 - val_loss: 0.4873
Epoch 6/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7544 - loss: 0.4961 - val_accuracy: 0.7594 - val_loss: 0.4840
Epoch 7/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7800 - loss: 0.4659 - val_accuracy: 0.7687 - val_loss: 0.4797
Epoch 8/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7725 - loss: 0.4665 - val_accuracy: 0.7634 - 

Using oversampling F1_score for class_1 increasing from 0.52 to 0.75

Using SMOTE

In [51]:
from imblearn.over_sampling import SMOTE
smote = SMOTE(sampling_strategy='minority')
X_sample,y_sample = smote.fit_resample(X,y)
y_sample.value_counts()

Churn
0    1869
1    1869
Name: count, dtype: int64

In [52]:
X_train,X_test,y_train,y_test = train_test_split(X_sample,y_sample,test_size=0.2,random_state=42,stratify=y_sample)
y_prediction = ANN(X_train,X_test,y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 4s 15ms/step - accuracy: 0.6715 - loss: 0.6195 - val_accuracy: 0.7273 - val_loss: 0.5386
Epoch 2/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7632 - loss: 0.4844 - val_accuracy: 0.7393 - val_loss: 0.5181
Epoch 3/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7695 - loss: 0.4764 - val_accuracy: 0.7393 - val_loss: 0.5141
Epoch 4/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7760 - loss: 0.4749 - val_accuracy: 0.7406 - val_loss: 0.5155
Epoch 5/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7791 - loss: 0.4744 - val_accuracy: 0.7473 - val_loss: 0.5160
Epoch 6/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7803 - loss: 0.4637 - val_accuracy: 0.7433 - val_loss: 0.5111
Epoch 7/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7885 - loss: 0.4538 - val_accuracy: 0.7433 - val_loss: 0.5098
Epoch 8/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7875 - loss: 0.4557 - val_accuracy: 0.7286 - 

Using SMOTE F1_score for class_1 increasing from 0.52 to 0.69

In [78]:
X_train,X_test,y_train,y_test = train_test_split(dt.drop(columns=['Churn']),dt['Churn'],test_size=0.2,random_state=42)

In [79]:
dt2 = X_train.copy()

In [80]:
dt2['Churn'] = y_train

In [81]:
dt2['Churn'].value_counts()

Churn
0    4138
1    1496
Name: count, dtype: int64

Using Ensemble Method ## divide into batches.

In [82]:
class_0 = dt2[dt2['Churn']==0]
class_1 = dt2[dt2['Churn']==1]
class_1.shape

(1496, 24)

In [83]:
4138/1496

2.766042780748663

In [84]:
1496*2

2992

In [85]:
def get_ensemble_batch(minority_dt,majority_dt,start,end):
  batch_dt = majority_dt.iloc[start:end]
  ensemble_batch_dt = pd.concat([batch_dt,minority_dt],axis=0)
  X_train = ensemble_batch_dt.drop(columns=['Churn'])
  y_train = ensemble_batch_dt['Churn']
  return X_train,y_train

In [86]:
X_train,y_train = get_ensemble_batch(class_1,class_0,0,1496)
y_prediction = ANN(X_train,X_test,y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - accuracy: 0.6560 - loss: 0.6311 - val_accuracy: 0.7793 - val_loss: 0.4677
Epoch 2/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7536 - loss: 0.5068 - val_accuracy: 0.7857 - val_loss: 0.4506
Epoch 3/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7578 - loss: 0.5048 - val_accuracy: 0.7771 - val_loss: 0.4514
Epoch 4/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7547 - loss: 0.4880 - val_accuracy: 0.7729 - val_loss: 0.4664
Epoch 5/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7665 - loss: 0.4899 - val_accuracy: 0.7708 - val_loss: 0.4570
Epoch 6/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.7660 - loss: 0.4864 - val_accuracy: 0.7700 - val_loss: 0.4578
Epoch 7/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7645 - loss: 0.4745 - val_accuracy: 0.7339 - val_loss: 0.4988
Epoch 8/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7736 - loss: 0.4837 - val_accuracy: 0.7410 - 

In [87]:
X_train,y_train = get_ensemble_batch(class_1,class_0,1497,2992)
y_prediction1 = ANN(X_train,X_test,y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 4s 15ms/step - accuracy: 0.6564 - loss: 0.6286 - val_accuracy: 0.7559 - val_loss: 0.5153
Epoch 2/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7628 - loss: 0.5094 - val_accuracy: 0.7587 - val_loss: 0.4948
Epoch 3/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7532 - loss: 0.5043 - val_accuracy: 0.7566 - val_loss: 0.4853
Epoch 4/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7581 - loss: 0.4939 - val_accuracy: 0.7594 - val_loss: 0.4742
Epoch 5/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7697 - loss: 0.4796 - val_accuracy: 0.7495 - val_loss: 0.4877
Epoch 6/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7709 - loss: 0.4695 - val_accuracy: 0.7473 - val_loss: 0.4900
Epoch 7/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7715 - loss: 0.4808 - val_accuracy: 0.7509 - val_loss: 0.4842
Epoch 8/100
94/94 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.7775 - loss: 0.4697 - val_accuracy: 0.7495 - 

Using ensemble F1_score for class_1 increasing from 0.52 to (0.62+0.59)/2

Oversampling method work well out of these four different methods.
It gives f1_score from 0.52 to 0.75